source/agora/test/EnrollmentManager.d

```diff
@@ -221,3 +221,188 @@ unittest
+
+            // note: we are not passing the previous tx set as we don't really
+            // need it at this point (might later be necessary for chain upgrades)
+            auto slot_idx = this.ledger.getBlockHeight() + 1;
+            this.nominate(slot_idx, data);
+        }
+    }
+
+    static class MisbehavingValidator : TestValidatorNode
+    {
+        /// Ctor
+        public this (Config config, Registry* reg, immutable(Block)[] blocks)
+        {
+            super(config, reg, blocks);
+        }
+
+        ///
+        protected override Nominator getNominator (
+            NetworkManager network, KeyPair key_pair, Ledger ledger,
+            TaskManager taskman)
+        {
+            return new BadNominator(network, key_pair, ledger, taskman);
+        }
+    }
+
+    static class OneTimeNominator : Nominator
+    {
+        private Ledger ledger;
+        private bool is_nominating;
+
+        /// Ctor
+        public this (NetworkManager network, KeyPair key_pair,
+            Ledger ledger, TaskManager taskman)
+        {
+            this.ledger = ledger;
+            super(network, key_pair, ledger, taskman);
+        }
+
+        ///
+        public override bool isNominating () @safe @nogc nothrow
+        {
+            return this.is_nominating;
+        }
+
+        ///
+        public override void tryNominate () @safe
+        {
+            if (this.ledger.getBlockHeight < 1)
+            {
+                super.tryNominate();
+            }
+        }
+    }
+
+    static class OneTimeValidator : TestValidatorNode
+    {
+        /// Ctor
+        public this (Config config, Registry* reg, immutable(Block)[] blocks)
+        {
+            super(config, reg, blocks);
+        }
+
+        ///
+        protected override Nominator getNominator (
+            NetworkManager network, KeyPair key_pair, Ledger ledger,
+            TaskManager taskman)
+        {
+            return new OneTimeNominator(network, key_pair, ledger, taskman);
+        }
+    }
+
+    static class BadAPIManager : TestAPIManager
+    {
+        ///
+        public this (immutable(Block)[] blocks)
+        {
+            super(blocks);
+        }
+
+        /// see base class
+        public override void createNewNode (Config conf)
+        {
+            if (this.nodes.length == 0)
+            {
+                auto api = RemoteAPI!TestAPI.spawn!MisbehavingValidator(
+                    conf, &this.reg, this.blocks, conf.node.timeout);
+                this.reg.register(conf.node.address, api.tid());
+                this.nodes ~= NodePair(conf.node.address, api);
+            }
+            else
+            {
+                auto api = RemoteAPI!TestAPI.spawn!OneTimeValidator(
+                    conf, &this.reg, this.blocks, conf.node.timeout);
+                this.reg.register(conf.node.address, api.tid());
+                this.nodes ~= NodePair(conf.node.address, api);
+            }
+        }
```

Comment:
Why do we need the `OneTimeValidator` ? I thought injecting a single node would be enough.

Reply:
> Why do we need the OneTimeValidator ? I thought injecting a single node would be enough.

I just wanted to make the case when the block won't be made if there is only a bad nominator. But, It may be more natural to have the situation where the nomination protocol weeds bad data out as you suggested.

Reply:
Yes, the issue was purely about rejecting that block at *nomination* time. Making it to a block would be a grave violation of consensus.